File: pkg/timestream/macros.go
package timestream

import (
	"fmt"
	"slices"
	"strings"
	"time"

	"github.com/grafana/grafana-plugin-sdk-go/experimental/errorsource"
	"github.com/grafana/timestream-datasource/pkg/models"
	"golang.org/x/exp/maps"
)

// TODO: consider refactoring sqlutil.Interpolate to be more generic and using that instead

type macroFunc func(models.QueryModel, models.DatasourceSettings) (string, error)

var macroFuncs = map[string]macroFunc{
	"timeFilter":      macroTimeFilter,
	"timeFrom":        macroTimeFrom,
	"timeTo":          macroTimeTo,
	"interval":        macroInterval,
	"interval_ms":     macroInterval,
	"interval_raw_ms": macroIntervalRaw,
	"now_ms":          macroNow,
	"database":        macroDatabase,
	"table":           macroTable,
	"measure":         macroMeasure,
}

var macroKeys []string

func init() {
	// sort macro keys longest first, so shorter keys don't clobber longer keys
	// they're a prefix of
	macroKeys = maps.Keys(macroFuncs)
	slices.SortFunc(macroKeys, func(a, b string) int { return len(b) - len(a) })
}

func macroTimeFilter(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
	from := model.TimeRange.From.UnixNano() / 1e6
	to := model.TimeRange.To.UnixNano() / 1e6

	replacement := fmt.Sprintf("time BETWEEN from_milliseconds(%d) AND from_milliseconds(%d)", from, to)
	return replacement, nil
}

func macroTimeFrom(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
	return fmt.Sprintf("%d", model.TimeRange.From.UnixNano()/1e6), nil
}

func macroTimeTo(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
	return fmt.Sprintf("%d", model.TimeRange.To.UnixNano()/1e6), nil
}

func macroInterval(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
	if model.Interval.Milliseconds() == 0 {
		return "", fmt.Errorf("invalid interval: %dns", model.Interval.Nanoseconds())
	}
	return fmt.Sprintf("%dms", model.Interval.Milliseconds()), nil
}

func macroIntervalRaw(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
	if model.Interval.Milliseconds() == 0 {
		return "", fmt.Errorf("invalid interval: %dns", model.Interval.Nanoseconds())
	}
	return fmt.Sprintf("%d", model.Interval.Milliseconds()), nil
}

func macroNow(_ models.QueryModel, _ models.DatasourceSettings) (string, error) {
	now := time.Now().UnixMilli()
	return fmt.Sprintf("%d", now), nil
}

func macroDatabase(model models.QueryModel, settings models.DatasourceSettings) (string, error) {
	return valueOrDefault(model.Database, settings.DefaultDatabase), nil
}
func macroTable(model models.QueryModel, settings models.DatasourceSettings) (string, error) {
	return valueOrDefault(model.Table, settings.DefaultTable), nil
}
func macroMeasure(model models.QueryModel, settings models.DatasourceSettings) (string, error) {
	return valueOrDefault(model.Measure, settings.DefaultMeasure), nil
}

func valueOrDefault(value string, defaultValue string) string {
	if value == "" || strings.HasPrefix(value, "${") {
		return defaultValue
	}
	return value
}

// Interpolate processes macros
func Interpolate(model models.QueryModel, settings models.DatasourceSettings) (string, error) {
	query := model.RawQuery
	for _, key := range macroKeys {
		macroKey := fmt.Sprintf("$__%s", key)
		if !strings.Contains(query, macroKey) {
			continue
		}
		replacement, err := macroFuncs[key](model, settings)
		if err != nil {
			return query, errorsource.DownstreamError(err, false)
		}
		query = strings.ReplaceAll(query, macroKey, replacement)
	}
	return query, nil
}

File: src/language/macros.ts
import { MacroType } from '@grafana/experimental';

export const TABLE_MACRO = '$__table';
export const DATABASE_MACRO = '$__database';

export const MACROS = [
  {
    id: '$__timeFilter',
    name: '$__timeFilter',
    text: '$__timeFilter',
    args: [],
    type: MacroType.Filter,
    description: 'Will be replaced by an expression that limits the time to the dashboard range.',
  },
  {
    id: '$__timeFrom',
    name: '$__timeFrom',
    text: '$__timeFrom',
    args: [],
    type: MacroType.Filter,
    description: 'Will be replaced by the number in milliseconds at the start of the dashboard range',
  },
  {
    id: '$__timeTo',
    name: '$__timeTo',
    text: '$__timeTo',
    args: [],
    type: MacroType.Filter,
    description: 'Will be replaced by the number in milliseconds at the end of the dashboard range.',
  },
  {
    id: '$__interval_ms',
    name: '$__interval_ms',
    text: '$__interval_ms',
    args: [],
    type: MacroType.Filter,
    description:
      'Will be replaced by a number in time format that represents the amount of time a single pixel in the graph should cover.',
  },
  {
    id: '$__interval_raw_ms',
    name: '$__interval_raw_ms',
    text: '$__interval_raw_ms',
    args: [],
    type: MacroType.Filter,
    description:
      'Will be replaced by the number in milliseconds that represents the amount of time a single pixel in the graph should cover.',
  },
  {
    id: DATABASE_MACRO,
    name: DATABASE_MACRO,
    text: DATABASE_MACRO,
    args: [],
    type: MacroType.Table,
    description: 'Will be replaced by the query database.',
  },
  {
    id: TABLE_MACRO,
    name: TABLE_MACRO,
    text: TABLE_MACRO,
    args: [],
    type: MacroType.Table,
    description: 'Will be replaced by the query table.',
  },
  {
    id: '$__measure',
    name: '$__measure',
    text: '$__measure',
    args: [],
    type: MacroType.Column,
    description: 'Will be replaced by the query measure.',
  },
];
